R/get_kr_comp.R

```diff
@@ -0,0 +1,115 @@
+#' obtain kr component
+#' @param tmb_data (`mmrm_tmb_data`)\cr produced by [h_mmrm_tmb_data()].
+#' @param theta numeric vector of the theta estimate.
+#' @details the function retuns a named list, $P$, $Q$ and $R$, which corresponds to the
+#' paper in 1997. The matrices are stacked in columns so that $P$, $Q$ and $R$ has the same
+#' column number(number of beta parameters). The number of rows, is dependent on
+#' the total number of theta and number of groups, if the fit is a grouped mmrm.
+#' For $P$ matrix, it is stacked sequentially. For $Q$ and $R$ matrix, it is stacked so
+#' that the $Q_{ij}$ and $R_{ij}$ is stacked from $j$ then to $i$, i.e. $R_{i1}$, $R_{i2}$, etc.
+#' $Q$ and $R$ only contains within group results and intra-group results should be all zero matrices
+#' so they are not stacked in the result.
+#' @keywords internal
+h_get_kr_comp <- function(tmb_data, theta) {
+  .Call(`_mmrm_get_pqr`, PACKAGE = "mmrm", tmb_data, theta)
+}
+
+
+#' obtain Kenward-Roger degree of freedom
+#' @export
+#' @param fit `mmrm` fit
+#' @param contrast contrast vector (ncol = 1)
+#' @examples
+#' fit <- mmrm(FEV1 ~ ARMCD + ar1(AVISIT | USUBJID), data = fev_data)
+#' contrast <- matrix(c(0,1), ncol = 1)
+#' kr(fit, contrast)
+kr <- function(fit, contrast, order = 2L) {
+  if (fit$tmb_data$reml != 1) {
+    stop("Kenward-Roger is only for REML!")
+  }
+  kr_comp <- fit$kr_comp
+  w <- solve(fit$tmb_obj$he(fit$theta_est))
+  v_adj <- v_a(fit$beta_vcov, w, kr_comp$P, kr_comp$Q, kr_comp$R, order = order)
+  df <- h_kr_df(fit$beta_vcov, contrast, w, kr_comp$P)
+  f_statistic <- 1 / nrow(contrast) * fit$beta_est %*% t(contrast) %*% solve(contrast %*% v_adj %*% t(contrast)) %*% contrast %*% fit$beta_est
+  f_star <- f_statistic * df$m / (df$m - 1 + nrow(contrast))
+  return(list(df = df, v_adj = v_adj, f_star = f_star))
+}
+
+#' obtain the adjusted Kenward-Roger degree of freedom
+#' @param v0 unadjusted covariance matrix
+#' @param l linear combination matrix
+#' @param w hessian matrix
+#' @param p P matrix from `h_get_kr_comp`
```

Comment:
even for internal functions please have `@return` documented and follow conventions for roxygen, thanks